### PR TITLE
Fix num. of minimal calls to the Hub with peft for pipeline 

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

This is a continuation of #25715 where @sgugger fix `test_cached_model_has_minimum_calls_to_head` but not `test_cached_pipeline_has_minimum_calls_to_head` (as I didn't mention this to him).

This PR mostly copies the logic from #25715.

The current failing test and the error are

```bash
tests/pipelines/test_pipelines_common.py::CustomPipelineTest::test_cached_pipeline_has_minimum_calls_to_head
(line 905)  AssertionError: 2 != 1
```

### CODE REVIEW

File: src/transformers/pipelines/__init__.py [hunk 1]
```diff
@@ -727,28 +733,48 @@ def pipeline(
     if isinstance(model, Path):
         model = str(model)
 
+    if hub_kwargs["_commit_hash"] is None:
+        pretrained_model_name_or_path = None
+        if isinstance(config, str):
+            pretrained_model_name_or_path = config
+        elif config is None and isinstance(model, str):
+            pretrained_model_name_or_path = model
```
Comment (ydshieh) [inline src/transformers/pipelines/__init__.py:741]: try to get the repo. in different case before going to the next block of getting config, similar to the block in line 758

File: src/transformers/pipelines/__init__.py [hunk 2]
```diff
@@ -769,7 +795,7 @@ def pipeline(
                 "Inferring the task automatically requires to check the hub with a model_id defined as a `str`."
                 f"{model} is not a valid model_id."
             )
-        task = get_task(model, use_auth_token)
+        task = get_task(model, token)
```
Comment (ydshieh) [inline src/transformers/pipelines/__init__.py:798]: `use_auth_token` is already deprecated.

File: src/transformers/pipelines/__init__.py [hunk 3]
```diff
@@ -727,28 +733,48 @@ def pipeline(
     if isinstance(model, Path):
         model = str(model)
 
+    if hub_kwargs["_commit_hash"] is None:
```
Comment (LysandreJik) [inline src/transformers/pipelines/__init__.py:736]: You have isolated it in a separate variable so you could just do 
```suggestion
    if commit_hash is None:
```

File: src/transformers/pipelines/__init__.py [hunk 4]
```diff
@@ -727,28 +733,48 @@ def pipeline(
...
         if is_peft_available():
-            subfolder = hub_kwargs.get("subfolder", None)
-            maybe_adapter_path = find_adapter_config_file(
-                model,
-                revision=revision,
-                token=use_auth_token,
-                subfolder=subfolder,
-            )
+            # `find_adapter_config_file` doesn't accept `trust_remote_code`
+            _hub_kwargs = {k: v for k, v in hub_kwargs.items() if k != "trust_remote_code"}
```
Comment (LysandreJik) [inline src/transformers/pipelines/__init__.py:767]: Can we just pass explicitly the arguments as it was done before? This seems a bit brittle

**Review by LysandreJik — APPROVED** [review]
This seems okay to me generally, would like another review as it's touching critical code.

Not too sure how we could do it, but it would also be nice to have tests for this, as it's the kind of thing that will continue to break unless we actively test it.

File: src/transformers/pipelines/__init__.py (hunk 4 above)
Comment (ydshieh) [reply src/transformers/pipelines/__init__.py:767]: Look this part 

https://github.com/huggingface/transformers/blob/0ac3875011d32dc85e0e83970507e3afe8f0febb/src/transformers/modeling_utils.py#L2511-L2522 (this [diff in modeling_utils.py](https://github.com/huggingface/transformers/commit/2febd506149d039b51590f5dc7b45f0d8624819d) might be more clear)

it passes more arguments to `find_adapter_config_file`. I can definitely use (a long list of) explicitly arguments here.

File: src/transformers/pipelines/__init__.py (hunk 3 above)
Comment (ydshieh) [reply src/transformers/pipelines/__init__.py:736]: That variable is mainly to be used to update

```bash
    hub_kwargs = {
        ....
        "_commit_hash": commit_hash,
    }
```
and anything after is best to continue to use `hub_kwargs["_commit_hash"]`.

I can avoid the definition of `commit_hash` and just use `hub_kwargs["_commit_hash"]` everywhere.

**Comment by ydshieh** [comment]
> This seems okay to me generally, would like another review as it's touching critical code.
> 
> Not too sure how we could do it, but it would also be nice to have tests for this, as it's the kind of thing that will continue to break unless we actively test it.

We do have tests: 4 tests named `test_cached_xxx_has_minimum_calls_to_head`. This is why I found the failure to fix.

File: src/transformers/pipelines/__init__.py (hunk 3 above)
Comment (ydshieh) [reply src/transformers/pipelines/__init__.py]: done locally and pushed 😅

File: src/transformers/pipelines/__init__.py [hunk 5]
```diff
@@ -727,28 +733,53 @@ def pipeline(
...
     if isinstance(config, str):
-        config = AutoConfig.from_pretrained(config, _from_pipeline=task, **hub_kwargs, **model_kwargs)
+        config = AutoConfig.from_pretrained(
+            config, _from_pipeline=task, code_revision=code_revision, **hub_kwargs, **model_kwargs
+        )
         hub_kwargs["_commit_hash"] = config._commit_hash
     elif config is None and isinstance(model, str):
         # Check for an adapter file in the model path if PEFT is available
         if is_peft_available():
-            subfolder = hub_kwargs.get("subfolder", None)
```
Comment (ydshieh) [inline src/transformers/pipelines/__init__.py:738]: `subfolder` was never in `hub_kwargs` here

**Review by sgugger — APPROVED** [review]
This looks good to me, thanks for expanding my fix!

**Comment by ydshieh** [comment]
Merge now.

**Review by ArthurZucker — APPROVED** [review]
Feel free to merge!

**Pull request merged**